_posts/2019-07-11-JS-testing-101.md

```diff
@@ -0,0 +1,433 @@
+
+## Asynchronous testing
+
+Thinking in async code can be a bit of a challenge, especially when you start out with JavaScript as a new developer or coming from a language that works mostly with synchronous code.
+
+In this part of the post, I will show you how to test asynchronous code. Both with promises and observables.
+
+For both promises and observables, I will test methods that show the most common usage patterns. One scenario is where you simply get something asynchronously and set its value to a field, the other scenario is a little more complicated, where you get something asynchronously, have some side effects before it resolves, and have some more side effects after it resolves. If that seems complicated, don’t worry, it will become clear very soon.
+For the promises, I will be using the async / await syntax. If you don’t know what that is yet, I suggest you read up on it, a good article is
+<a href="https://alligator.io/js/async-functions/">this one</a>.
+
+### Promises
+
+#### Scenario A
+
+ The method we will be testing will be the following:
+ 
+````javascript
+async getAllRedCarsAsPromise(): Promise<Array<Car>> {
+    const redCars = await this.carService.getCarsAsPromise();
+    return redCars.filter((car: Car) => car.color === 'red');
+}
+````
+ 
+This function just gets the cars from the `carService` asynchronously and filters out all the non-red cars before returning the array. 
+ 
+The first test, we will simply test the happy scenario, where the `carService` returns us a nice list of cars and the non-red cars are filtered out.
+
+````javascript
+ describe('getAllRedCarsAsPromise', () => {
+
+    it('should get the cars where the color is red', async done => {
+      when(carService.getCarsAsPromise()).thenReturn(Promise.resolve([aRedPorsche, aBlackCadillac, aBrownCar]));
+
+      try {
+        const actual = await componentUnderTest.getAllRedCarsAsPromise();
+
+        expect(actual).toEqual([aRedPorsche]);
+      } catch (e) {
+        fail(e);
+      } finally {
+   done();
+ }
+    });
+});
+````
+
+So what exactly goes on here? First we mock the carService.getCarsAsPromise() method and tell it that it should always return a promise that resolves with a list of cars. Then we call the method that we are testing, we wait for it to resolve (by using the await keyword). Afterward we expect that the returned array only contains the red Porsche. Now, what is this whole try catch thingy? When the promise is rejected, the execution of the code in the try block will be stopped and the code in the catch block will be executed. In this case, if the getCarsAsPromise() should fail, it will go into the catch block and the test will fail. In the finally block, we call the done function, to tell jasmine that our test is done, we call it in the finally block, because the code in the finally block will always be executed, regardless of the result of the promise. 
```

Comment:
The test will fail, but what will `fail(e)` output? Handy for the user to already know they can also add a string as shown below in the next example?